src/plugins/discover/public/application/helpers/get_field_list.ts

```diff
@@ -0,0 +1,22 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * and the Server Side Public License, v 1; you may not use this file except in
+ * compliance with, at your election, the Elastic License or the Server Side
+ * Public License, v 1.
+ */
+import { IndexPattern } from '../../../../data/common/index_patterns/index_patterns';
+
+const META_FIELDS = ['_type', '_source'];
+
+export function getFieldListFromIndexPattern(indexPattern: IndexPattern) {
+  const { sourceFilters, fields } = indexPattern;
+  if (!sourceFilters || sourceFilters.length === 0) {
+    return ['*'];
+  }
+  const sourceFiltersValues = sourceFilters.map((sourceFilter) => sourceFilter.value);
+  const fieldsToInclude = fields.filter((field) => {
+    return !sourceFiltersValues.includes(field.name) && !META_FIELDS.includes(field.name);
+  });
```

Comment:
I may be overlooking something (and things might have changed since I last looked at this stuff), but I'm not sure if this part should be necessary.


The idea with search source is that you provide the index pattern, and this stuff happens internally so you don't need to worry about it. The filtering (and meta fields handling) should already happen inside search source via `fieldsWildcardFilter`.

As a consumer of search source, all you should be responsible for is providing a field list, and search source would ideally handle everything else.

But I might be oversimplifying: Is this addressing a case that is not already covered by search source?

Reply:
So this particular use case is supposed to take care of `field filters` when using the fields API. Unlike `_source`, fields API has no notion of `excludes`, but rather - `includes`. So we'd have to request a list of all fields, except the ones specified by the filters, which is what this code does. The logic doesn't exist (or doesn't work 😅 ) in `search_source` yet. TBF, I am not sure where the right place for it is. It felt cleaner to be in Discover, but happy to discuss alternatives.

Reply:
> fields API has no notion of excludes, but rather - includes. So we'd have to request a list of all fields, except the ones specified by the filters, which is what this code does.

Ah, this is the point I missed. Yeah this does ultimately feel like something search source should handle. I would expect that if you called `setField('fields', ['*'])`, then search source would do this internally. But let's see what @ppisljar thinks

Reply:
After some discussions on `date_nanos` handling, we have reached the conclusion that Discover would explicitly specify the date format for any date field when requesting fields (https://github.com/elastic/kibana/pull/90415#issuecomment-774158585). I believe this logic belongs to the same place as requesting fields to include, so if we could now decide where that would be, it would be appreciated :)

Reply:
Current functionality was that when not setting the fields we would get you all of them. We changed this on the searchSource level, so that we would no longer return unmapped ones when using the fieldsAPI. I think there is place for this logic to live in the searchsource, thus get to the old behaviour. We could also make this optional by exposing `showUnmappedFields' on the searchsource.

it seems from the linked PR that we are handling date_nanos on searchsource level as well, which makes sense imo

Reply:
I will change this then so that the logic lives in search source 👍